libavcodec/mfdec.c

```diff
@@ -0,0 +1,1717 @@
+            return AVERROR_EXTERNAL;
+        }
+
+        if (c->opt_d3d_bind_flags >= 0) {
+            hr = IMFAttributes_SetUINT32(attrs, &ff_MF_SA_D3D11_BINDFLAGS, c->opt_d3d_bind_flags);
+            if (FAILED(hr))
+                av_log(avctx, AV_LOG_ERROR, "could not set bindflags (%s)\n", ff_hr_str(hr));
+        }
+
+        IMFAttributes_Release(attrs);
+    }
+
+    return 0;
+}
+
+static LONG mf_codecapi_get_int(ICodecAPI *capi, const GUID *guid, LONG def)
+{
+    LONG ret = def;
+    VARIANT v;
+    HRESULT hr = ICodecAPI_GetValue(capi, &ff_CODECAPI_AVDecVideoMaxCodedWidth, &v);
+    if (FAILED(hr))
+        return ret;
+    if (v.vt == VT_I4)
+        ret = v.lVal;
+    if (v.vt == VT_UI4)
+        ret = v.ulVal;
+    VariantClear(&v);
+    return ret;
+}
+
+static int mf_check_codec_requirements(AVCodecContext *avctx)
+{
+    MFContext *c = avctx->priv_data;
+
+    if (c->is_dec && c->is_video && c->codec_api) {
+        LONG w = mf_codecapi_get_int(c->codec_api, &ff_CODECAPI_AVDecVideoMaxCodedWidth, 0);
+        LONG h = mf_codecapi_get_int(c->codec_api, &ff_CODECAPI_AVDecVideoMaxCodedHeight, 0);
+
+        if (w <= 0 || h <= 0)
+            return 0;
+
+        av_log(avctx, AV_LOG_VERBOSE, "Max. supported video size: %dx%d\n", (int)w, (int)h);
+
+        // avctx generally has only the cropped size. Assume the coded size is
+        // the same size, rounded up to the next macroblock boundary.
+        if (avctx->width > w || avctx->height > h) {
+            av_log(avctx, AV_LOG_ERROR, "Video size %dx%d larger than supported size.\n",
+                   avctx->width, avctx->height);
+            return AVERROR(EINVAL);
+        }
+    }
+
+    return 0;
+}
+
+static int mf_unlock_async(AVCodecContext *avctx)
+{
+    // decoding does not neex async event handling
+    return 0;
+}
+
+static void mf_release_decoder(void *opaque, uint8_t *data)
+{
+    MFDecoder *dec = (void *)data;
+
+    // At least async MFTs require this to be called to truly terminate it.
+    // Of course, mingw is missing both the import lib stub for
+    // MFShutdownObject, as well as the entire IMFShutdown interface.
+    HANDLE lib = LoadLibraryW(L"mf.dll");
+    if (lib) {
+        HRESULT (WINAPI *MFShutdownObject_ptr)(IUnknown *pUnk)
+            = (void *)GetProcAddress(lib, "MFShutdownObject");
+        if (MFShutdownObject_ptr)
+            MFShutdownObject_ptr((IUnknown *)dec->mft);
+        FreeLibrary(lib);
+    }
+
+    IMFTransform_Release(dec->mft);
+
+    av_buffer_unref(&dec->device_ref);
+}
+
+static int mf_init(AVCodecContext *avctx)
+{
+    MFContext *c = avctx->priv_data;
+    HRESULT hr;
+    int ret;
+    MFDecoder *dec;
+    const CLSID *subtype = ff_codec_to_mf_subtype(avctx->codec_id);
+    int use_hw = 0;
+
+    c->tmp_frame = av_frame_alloc();
+    if (!c->tmp_frame)
+        return AVERROR(ENOMEM);
+
+    c->original_channels = avctx->channels;
+
+    c->is_dec = av_codec_is_decoder(avctx->codec);
```

Comment:
Related to the `is_enc` comment, but could this potentially just error out if `is_dec` is false? That way we can avoid verifying it in other spots in this code

Reply:
Cleaned up unneeded is_dec usage.